### PR TITLE
Updating Documentation for e2e granite profiling

### LINKED ISSUE #1963: Validate end-to-end profiling example on Spyre hardware and update docs
## Background

PR #1954 added [`docs/source/user_guide/profiling/end_to_end_example.md`](https://github.com/torch-spyre/torch-spyre/blob/main/docs/source/user_guide/profiling/end_to_end_example.md): an end-to-end recipe for capturing a `torch.profiler` trace of a Granite-class model on Spyre, paired with `aiu-smi` device telemetry.

The page is under review with the test plan item *"Validate the example end-to-end on Spyre hardware"* still unchecked.

## What to do

1. Run the example end-to-end on Spyre hardware following the page as written:
   - Install the listed packages (`foundation-model-stack`, `aiu-fms-testing-utils`, `kineto-spyre`, optionally `aiu-trace-analyzer`).
   - Download the Granite checkpoint.
   - Run `profile_granite.py`.
   - Run `aiu-smi dmon` alongside (with `SENLIB_DEVEL_CONFIG_FILE` set).
   - Open the resulting trace in Perfetto / TensorBoard.
2. Note any drift between the page and reality: wrong env var names, missing setup steps, command-line flags that have changed, output that doesn't match the description, etc.
3. Add screen shots showing examples 
3. Open a follow-up PR with the corrections.

## Things to specifically verify

- [ ] `fms.get_model(...)` arguments are still current (`architecture`, `device_type`, `data_type`, `unfuse_weights`).
- [ ] `model.compile()` invocation works as written — or should be `torch.compile(model)`.
- [ ] `ProfilerActivity.PrivateUse1` produces device-side rows in `prof.key_averages()`.
- [ ] `sort_by="device_time_total"` is the right column name (the registered device is `spyre`, so no `CUDA→AIU` rename should be needed).
- [ ] The `kineto-spyre` wheel URL example still resolves; pick a version that matches the current PyTorch in the dev environment.
- [ ] `SENLIB_DEVEL_CONFIG_FILE` is the correct env var name (or substitute the right one).
- [ ] `aiu-smi dmon` columns match what the page describes (power, temperature, PT-array util, device memory, PCIe BW).
- [ ] The "What to look for" table signals are observable in a real run.

## Deliverable

A follow-up PR that updates [`end_to_end_example.md`](https://github.com/torch-spyre/torch-spyre/blob/main/docs/source/user_guide/profiling/end_to_end_example.md) with any corrections and adds example output.

Originating PR: #1954

### PR DESCRIPTION
**Pull request by WinnieImafidon1**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [ ] feature
- [x ] documentation
- [ ] cleanup

#### Does this PR introduce a user-facing change?

Yes gives users a guide on how to use the granite model and profile.

### CODE REVIEW

File: docs/source/user_guide/profiling/end_to_end_example.md [hunk 1]
```diff
@@ -1,12 +1,10 @@
-# End-to-End Example: Profiling a Granite Model on Spyre
+# End-to-End Example: Profiling a Granite Model on Spyre via FMS
 
-**Stack:** torch-spyre (new, Inductor-based).
+**Stack:** Torch-Spyre (new, Inductor-based).
```
Comment (kaoutar55) [inline docs/source/user_guide/profiling/end_to_end_example.md]: Two things here. The project name is written `torch-spyre` (lowercase) everywhere else, so please keep it consistent. More importantly, the page now profiles in eager mode, but this line still describes the stack as "Inductor-based" and the admonition below still frames torch.compile as the path. Since compile has been removed from the script, reconcile the framing so it does not contradict the eager example.

File: docs/source/user_guide/profiling/end_to_end_example.md [hunk 2]
```diff
@@ -26,13 +26,22 @@ shrink. This page will be revised to the in-tree API once it ships.
 
 ## What you need
 
+**Prerequisites Step**
+
+Ensure that you have access a pod with spyre accelerator and torch spyre and spyre software stack installed.
+
+**Prerequisites Step**
```
Comment (kaoutar55) [inline docs/source/user_guide/profiling/end_to_end_example.md]: This "Prerequisites Step" block is duplicated (lines 29-31 and 33-35 are identical). Please keep one copy. Also a small grammar fix: "access a pod" should read "access to a pod", and "spyre"/"torch spyre" read better capitalized as "Spyre" / "torch-spyre" to match the rest of the page.

File: docs/source/user_guide/profiling/end_to_end_example.md [hunk 3]
```diff
@@ -26,13 +26,22 @@ shrink. This page will be revised to the in-tree API once it ships.
...
+| `kineto-spyre` (not required for torch >= 2.13)| [github.com/IBM/kineto-spyre][kineto-spyre] | `uv pip install --no-deps <release-wheel-url-matching-your-pytorch>` (see [releases page][kineto-spyre-releases]) |
 | `aiu-trace-analyzer` (optional) | [github.com/IBM/aiu-trace-analyzer][ata] | `pip install aiu-trace-analyzer` |
-| Granite checkpoint | [huggingface.co/ibm-granite/granite-3.3-8b-instruct](https://huggingface.co/ibm-granite/granite-3.3-8b-instruct) | `huggingface-cli download ibm-granite/granite-3.3-8b-instruct --local-dir /tmp/models/granite-3.3-8b-instruct` |
+| Granite checkpoint | [huggingface.co/ibm-granite/granite-3.3-8b-instruct](https://huggingface.co/ibm-granite/granite-3.3-8b-instruct) | `hf download ibm-granite/granite-3.3-8b-instruct --local-dir /tmp/models/granite-3.3-8b-instruct` |
+| Granite checkpoint | [huggingface.co/ibm-granite/granite-3.3-8b-instruct](https://huggingface.co/ibm-granite/granite-3.3-8b-instruct) | `hf download ibm-granite/granite-3.3-8b-instruct --local-dir /tmp/models/granite-3.3-8b-instruct` |
```
Comment (kaoutar55) [inline docs/source/user_guide/profiling/end_to_end_example.md]: The Granite checkpoint row is duplicated (this line repeats line 43). Remove the extra row.

File: docs/source/user_guide/profiling/end_to_end_example.md [hunk 4]
```diff
@@ -26,13 +26,22 @@ shrink. This page will be revised to the in-tree API once it ships.
...
+**Prerequisites Step**
+
+Ensure that you have access a pod with spyre accelerator and torch spyre and spyre software stack installed.
+
 | Piece | Source | Sample install (verify against the upstream README) |
 |---|---|---|
 | `foundation-model-stack` (`fms`) | [github.com/foundation-model-stack/foundation-model-stack][fms] (`eager_spyre` branch) | `git clone -b eager_spyre <repo>.git && uv pip install -e ./foundation-model-stack` |
 | `aiu-fms-testing-utils` | [github.com/foundation-model-stack/aiu-fms-testing-utils][aiu-fms] (`eager_spyre` branch) | `git clone -b eager_spyre <repo>.git && uv pip install -e ./aiu-fms-testing-utils` |
-| `kineto-spyre` | [github.com/IBM/kineto-spyre][kineto-spyre] | `uv pip install --no-deps <release-wheel-url-matching-your-pytorch>` (see [releases page][kineto-spyre-releases]) |
+| `kineto-spyre` (not required for torch >= 2.13)| [github.com/IBM/kineto-spyre][kineto-spyre] | `uv pip install --no-deps <release-wheel-url-matching-your-pytorch>` (see [releases page][kineto-spyre-releases]) |
```
Comment (kaoutar55) [inline docs/source/user_guide/profiling/end_to_end_example.md]: The kineto-spyre version threshold is stated inconsistently across the page. Here it says "not required for torch >= 2.13", the setup comment on line 71 says the same but then installs a 2.11 wheel, and the intro no longer lists kineto as part of the tooling. Please pick one threshold and state it the same way in all three places, and make clear that the 2.11 wheel is only an example for users still on 2.11.

File: docs/source/user_guide/profiling/end_to_end_example.md [hunk 5]
```diff
@@ -49,15 +58,21 @@ git clone -b eager_spyre https://github.com/foundation-model-stack/aiu-fms-testi
...
+
+#Download granite
+hf download ibm-granite/granite-3.3-8b-instruct --local-dir /tmp/models/granite-3.3-8b-instruct
+
+# If your environment uses PyTorch 2.11, install the matching kineto-spyre wheel. It is not required for PyTorch 2.13 and later
+# See below an example of 2.11 kineto Wheel on x86
 uv pip install --no-deps --force-reinstall \
-  https://github.com/IBM/kineto-spyre/releases/download/torch-2.10.0.aiu.kineto.1.1.1/torch-2.10.0+aiu.kineto.1.1.1-cp312-cp312-linux_x86_64.whl
+  https://github.com/IBM/kineto-spyre/releases/download/torch-2.11.0.aiu.kineto.1.1.2/torch-2.11.0+aiu.kineto.1.1.2-cp312-cp312-linux_x86_64.whl
+  https://github.com/IBM/kineto-spyre/releases/download/torch-2.11.0.aiu.kineto.1.1.2/torch-2.11.0+aiu.kineto.1.1.2-cp312-cp312-linux_x86_64.whl
```
Comment (kaoutar55) [inline docs/source/user_guide/profiling/end_to_end_example.md]: This install command is broken. The wheel URL is duplicated, and the first URL line has no trailing backslash, so the shell treats the second URL as a separate command and it will fail. Keep a single URL:

```bash
uv pip install --no-deps --force-reinstall \
  https://github.com/IBM/kineto-spyre/releases/download/torch-2.11.0.aiu.kineto.1.1.2/torch-2.11.0+aiu.kineto.1.1.2-cp312-cp312-linux_x86_64.whl
```

File: docs/source/user_guide/profiling/end_to_end_example.md [hunk 6]
```diff
@@ -77,43 +92,69 @@ export SENCORES=32                 # full accelerator (1–32; default 32)
...
+
+position_ids = kwargs["position_ids"]
+alpha = model.base_model.rot_emb.compute_freqs_cis(DEVICE, ids.shape[1])
+selected_freqs = model.base_model.rot_emb.cached_freqs[0][alpha][position_ids].to(DEVICE)
+mask = kwargs["mask"].to(dtype=torch.float16).to(DEVICE)
 
-# 4. Warm up — first call triggers torch.compile + kernel codegen.
-with torch.no_grad():
-    model(inputs)
+# 3Warmup — first run is always slower due to JIT/compilation
```
Comment (kaoutar55) [inline docs/source/user_guide/profiling/end_to_end_example.md]: Typo in the comment: `# 3Warmup` should read `# 3. Warm up`. Since the example is eager-only, "JIT/compilation" is also misleading here. The slowness is runtime and device initialization, not torch.compile, so please word it that way.

File: docs/source/user_guide/profiling/end_to_end_example.md [hunk 7]
```diff
@@ -77,43 +92,69 @@ export SENCORES=32                 # full accelerator (1–32; default 32)
...
+ids, kwargs = pad_input_ids(
+    [raw_ids],
+    min_pad_length=512,
+    pad_token_id=pad_token_id,
+)
+
+position_ids = kwargs["position_ids"]
+alpha = model.base_model.rot_emb.compute_freqs_cis(DEVICE, ids.shape[1])
+selected_freqs = model.base_model.rot_emb.cached_freqs[0][alpha][position_ids].to(DEVICE)
+mask = kwargs["mask"].to(dtype=torch.float16).to(DEVICE)
```
Comment (kaoutar55) [inline docs/source/user_guide/profiling/end_to_end_example.md:124]: This input preparation reaches into FMS internals (`model.base_model.rot_emb.compute_freqs_cis(...)`, `rot_emb.cached_freqs[0][alpha][position_ids]`) and passes `selected_freqs=` as a forward kwarg. Is this the intended path for eager Granite runs, or a temporary workaround? It is fragile against FMS refactors, and a reader copying it may hit breakage on a future FMS version. If it is required, a one-line note explaining why would help; if there is a public helper for this, prefer that.

File: docs/source/user_guide/profiling/end_to_end_example.md [hunk 8]
```diff
@@ -125,30 +166,36 @@ with profile(
     for _ in range(N_RUNS):
         t0 = time.perf_counter()
         with torch.no_grad():
-            model(inputs)
+          model(ids.to(DEVICE), position_ids=position_ids.to(DEVICE), mask=mask, selected_freqs=selected_freqs)
+          model(ids.to(DEVICE), position_ids=position_ids.to(DEVICE), mask=mask, selected_freqs=selected_freqs)
```
Comment (kaoutar55) [inline docs/source/user_guide/profiling/end_to_end_example.md]: The forward pass is called twice inside the timed loop, which corrupts the measurement this page is teaching. Each timed iteration now runs two forward passes, so `wall_clock_ms` records roughly double the per-run latency, while `cpu_per_run_ms` divides by `N_RUNS` assuming one call per iteration. Remove the duplicate call so there is a single `model(...)` per iteration.

File: docs/source/user_guide/profiling/end_to_end_example.md [hunk 9]
```diff
@@ -125,30 +166,36 @@ with profile(
...
 #    profiler-derived CPU time (host-side overhead). The gap between
 #    them ≈ device-side work.
 cpu_per_run_ms = sum(e.self_cpu_time_total for e in prof.events()) / 1000 / N_RUNS
 
+print("=" * 42)
+print("Profiling Granite Running on Spyre".center(42))
+print("=" * 42)
+print("=" * 42)
+print("Profiling Granite Running on Spyre".center(42))
+print("=" * 42)
```
Comment (kaoutar55) [inline docs/source/user_guide/profiling/end_to_end_example.md]: This "Profiling Granite Running on Spyre" print banner is duplicated (lines 179-181 repeat as 182-184). Keep one copy.

File: docs/source/user_guide/profiling/end_to_end_example.md [hunk 10]
```diff
@@ -198,7 +245,7 @@ For a Granite-class transformer the typical signals are:
 
 | Symptom | Likely cause | Where to dig |
 |---|---|---|
-| First iteration much slower than the rest | `torch.compile` warmup | Expected — discard the first iteration. |
+| First iteration much slower than the rest | the first iterations as runtime/device initialization or general warmup.
```
Comment (kaoutar55) [inline docs/source/user_guide/profiling/end_to_end_example.md]: This table row is broken: it lost the second and third columns and the closing pipe, so the whole table renders malformed. The sentence is also incomplete. Restore three cells, for example:

```
| First iteration much slower than the rest | Runtime and device initialization or general warmup | Expected. Discard the early iterations. |
```

File: docs/source/user_guide/profiling/end_to_end_example.md [hunk 11]
```diff
@@ -125,30 +166,36 @@ with profile(
...
-- **Run iteration 1 outside the timed loop.** First call carries the
-  `torch.compile` cost; including it skews the average.
-- **Two orthogonal timing signals.** Wall-clock from
-  `time.perf_counter()` is what the user feels; profiler-derived CPU
-  time (`sum(e.self_cpu_time_total)`) is host-side overhead. The
-  difference between them ≈ device-side work.
-- **`tensorboard_trace_handler(log_dir)` over `export_chrome_trace`.**
-  Per-step JSON files isolate compile-time warmup from steady-state
+- **Run warmup iterations outside the timed loop**. The first few executions can include one-time runtime initialization, lazy setup, cache population, or other startup overhead. Excluding them keeps steady-state measurements representative.
+- **Use two orthogonal timing signals**. Wall-clock time from time.perf_counter() represents end-to-end latency observed by the caller, while profiler-derived CPU time provides a view of host-side activity. Comparing them can help identify whether latency is dominated by host-side work or accelerator execution, but their difference should not be treated as a direct measurement of device time.
```
Comment (kaoutar55) [inline docs/source/user_guide/profiling/end_to_end_example.md]: These bullets are wordier and more hedged than the crisp versions they replace, and they run as single long lines while the rest of the file wraps at about 72 characters. Please rewrap to match and tighten the prose. Suggested:

- **Run warmup iterations outside the timed loop.** The first runs carry one-time runtime and device initialization. Excluding them keeps the steady-state numbers representative.
- **Read two timing signals together.** Wall-clock from `time.perf_counter()` is end-to-end latency; profiler-derived CPU time is host-side activity. Comparing them shows whether latency is host-bound or device-bound, though the difference is not a direct measurement of device time.

File: docs/source/user_guide/profiling/end_to_end_example.md [hunk 12]
```diff
@@ -125,30 +166,36 @@ with profile(
...
+print("Profiling Granite Running on Spyre".center(42))
+print("=" * 42)
 print(prof.key_averages().table(sort_by="device_time_total", row_limit=10))
 print(f"wall-clock ms: mean={mean(wall_clock_ms):.3f} median={median(wall_clock_ms):.3f}")
 print(f"profiler-derived CPU ms (per run): {cpu_per_run_ms:.3f}")
 ```
 
+**Note**
+
+Ensure the profiler report includes Spyre/device time; otherwise the trace does not confirm the accelerator execution
```
Comment (kaoutar55) [inline docs/source/user_guide/profiling/end_to_end_example.md]: Small wording fix for this note, and please wrap it to the file's line width: "Confirm the profiler table reports Spyre or device time. If it does not, the trace is not capturing accelerator execution."

**Review by kaoutar55 — CHANGES_REQUESTED** [review]
Thanks for taking this on. The direction is right: showing the eager Granite path via FMS is useful, and dropping the compile step for this page makes sense. A few things need cleanup before merge. Most look like rebase or copy-paste artifacts (several blocks are duplicated), and there are two correctness issues in the script and the setup command that would trip up anyone following along. Details inline.

One framing point that runs through the whole page: since the example is now eager-only, the intro still calls the stack "Inductor-based" and the admonition talks about torch.compile landing. Please align that framing so a reader does not get contradictory signals about whether compile is involved.

File: docs/source/user_guide/profiling/end_to_end_example.md (hunk 7 above)
Comment (WinnieImafidon1) [reply docs/source/user_guide/profiling/end_to_end_example.md:124]: Currently based on Antoni's suggestion this is a temporary work around once this is fixed we won't need this. I will update the documentation once this fix has landed

**Comment by kaoutar55** [comment]
@WinnieImafidon1 I pushed a small commit (b79c07a) to the branch that clears the remaining review items so we can merge:

- Restored the broken "first iteration slower" table row to three cells and filled the empty work-division cell in the "What to look for" table.
- Removed the orphaned "runs and open in TensorBoard..." fragment left after the tensorboard_trace_handler bullet.
- Dropped the unused `import os`, aligned the step-5 code comment with the corrected timing-signal wording, capitalized Spyre / torch-spyre in the prerequisites line, and added backticks to `torch.compile`.
- Added a one-line note marking the rotary-frequency input prep as a temporary FMS-internals workaround, per your reply on that thread.

pymarkdown passes and `make html` builds cleanly. Approving now. Thanks for the work on this.

**Review by kaoutar55 — APPROVED** [review]
Approving. The eager-mode FMS rework reads clearly, and the earlier correctness issues (double forward pass in the timed loop, broken install command, duplicated blocks) are resolved. The last cleanup items are handled in b79c07a. Thanks @WinnieImafidon1.

**Comment by kaoutar55** [comment]
@WinnieImafidon1 please change this PR from draft mode.

**Comment by kaoutar55** [comment]
@WinnieImafidon1 please find the unverified signature so we can merge this.
